compiler/rustc_mir_transform/src/simplify.rs

```diff
@@ -547,7 +547,10 @@ impl<'tcx> Visitor<'tcx> for UsedLocals {
     }
 
     fn visit_var_debug_info(&mut self, var_debug_info: &VarDebugInfo<'tcx>) {
-        if !self.preserve_debug && debug_info_is_for_simple_local(var_debug_info).is_some() {
+        // We don't want to have to track *conditional* uses (such as
+        // "`_4` is used iff `_5` is used" from `debug x => Foo(_4, _5)`),
+        // so if this mentions multiple locals we just treat them all as used.
```

Comment:
Should we just drop parts of the composite? If `_4` is unused, go from `Foo { .0 => _4, .1 => _5 }` to just `Foo { .1 => _5 }`?
This line can definitely just be a `if !self.preserve_debug`, and complexity will go in `remove_unused_definitions_helper`.